### PR TITLE
iter8-Lambda functions and Enums

### PR DESCRIPTION
**Pull request by jinocencio-solvd**
### Task - Lambda functions and Enums

Requirements

- Use at least 5 lambda functions from the java.util.function package.
    - `Predicate<T>` was used in `Pharmacy.isOpen()` to create a function that checks for the
      given `DayOfWeek` as a `BusinessDay` enum and whether the business is open on that day by
      checking the `BusinessDay` field `isOpen`. The `DayOfWeek` input is applied to
      the `Predicate.test` method and the result is returned as the output of the isOpen method.
    - `Consumer<T>` was used in `Main.prescriptionOperations()` to
      call `Pharmacist.fulfillAllPrescriptionLogRequests`. The function is subsequently used in the
      for loop to demonstrate max number of times for prescription refilling.
    - `Function<T,R>` was used in `Register.calculateDiscountPrice()` to transform and read
      an `AbstractCustomer` as a `Patient` during runtime.
    - `Supplier<T>` was used in `AbstractCustomer.increaseCreditBalanceByOneHundred()` to supply the
      patient with an increase of 100 to the customer's creditBalance.
    - `BooleanSupplier` is used in `AbstractCustomer.isPatient()` to return a boolean value if the
      method was invoked from the Patient class
- Create 3 custom Lambda functions with generics.
    - `customLambdaFunctions/PerformOperation` defines `PerformOperation<T>` that takes an object
      to `perform` an operation of that object.
    - `customLambdaFunctions/Repeater` defines `Repeater<T>` which
      invokes `repeat(int numRepeats, PerformOperation<T> operation)`. This lambda function repeats
      a `PerformOperation` by `numRepeats` amount of times.
    - customLambdaFunctions/NullChecker defines `NullChecker<T>` which returns a boolean value in
      regard to if the object passed into the invoke method `isNull` is a null value.
- Create 5 complex Enums(with fields, methods, blocks).
    - `misc/BusinessDays` include the days of the week and has `.id`, `.description`, and `.isOpen`
      fields. Methods include field getters and a `getBusinessDay()` method.
    - `prescriptionRegistry/PrescriptionStatus` include the various prescription states that are
      allowed. The status are set and modified by appropriate authorities i.e. initialized as `NEW`
      if unregistered, `PENDING` if received by pharmacy, `COMPLETED` and `REFILL_UPON_REQUEST` by
      pharmacist. `FILLED` will be implemented upon completion of patient transaction.
    - `product/ProductType` represent the different types of products that the pharmacy carries.
      All `Medications` are initialized with ProductType enum `PRESCRIBED_MEDICINE`. Products can be
      initialized upon instantiation or set with the setter method. Otherwise, all other products
      and items are `UNLABELED` by default.
    - `register/PaymentType` represents the different types of payment a customer can provide to
      complete their transaction at the register. Thus, `PaymentType` is a field implemented
      in `Register`. `NOT_SELECTED` is the default enum and can be changed by either the employee
      operating the register or the customer in future development.
    - `person/EmployeeType` represents the different types of employee the pharmacy can have. By
      default, all employees are `FULL_TIME` employees.

### CODE REVIEW

**Comment by jinocencio-solvd** [comment]
## Iter-8

### Task - Lambda functions and Enums

Requirements

- Use at least 5 lambda functions from the java.util.function package.
    - `Predicate<T>` was used in `Pharmacy.isOpen()` to create a function that checks for the
      given `DayOfWeek` as a `BusinessDay` enum and whether the business is open on that day by
      checking the `BusinessDay` field `isOpen`. The `DayOfWeek` input is applied to
      the `Predicate.test` method and the result is returned as the output of the isOpen method.
    - `Consumer<T>` was used in `Main.prescriptionOperations()` to
      call `Pharmacist.fulfillAllPrescriptionLogRequests`. The function is subsequently used in the
      for loop to demonstrate max number of times for prescription refilling.
    - `Function<T,R>` was used in `Register.calculateDiscountPrice()` to transform and read
      an `AbstractCustomer` as a `Patient` during runtime.
    - `Supplier<T>` was used in `AbstractCustomer.increaseCreditBalanceByOneHundred()` to supply the
      patient with an increase of 100 to the customer's creditBalance.
    - `BooleanSupplier` is used in `AbstractCustomer.isPatient()` to return a boolean value if the
      method was invoked from the Patient class
- Create 3 custom Lambda functions with generics.
    - `customLambdaFunctions/PerformOperation` defines `PerformOperation<T>` that takes an object
      to `perform` an operation of that object.
    - `customLambdaFunctions/Repeater` defines `Repeater<T>` which
      invokes `repeat(int numRepeats, PerformOperation<T> operation)`. This lambda function repeats
      a `PerformOperation` by `numRepeats` amount of times.
    - customLambdaFunctions/NullChecker defines `NullChecker<T>` which returns a boolean value in
      regard to if the object passed into the invoke method `isNull` is a null value.
- Create 5 complex Enums(with fields, methods, blocks).
    - `misc/BusinessDays` include the days of the week and has `.id`, `.description`, and `.isOpen`
      fields. Methods include field getters and a `getBusinessDay()` method.
    - `prescriptionRegistry/PrescriptionStatus` include the various prescription states that are
      allowed. The status are set and modified by appropriate authorities i.e. initialized as `NEW`
      if unregistered, `PENDING` if received by pharmacy, `COMPLETED` and `REFILL_UPON_REQUEST` by
      pharmacist. `FILLED` will be implemented upon completion of patient transaction.
    - `product/ProductType` represent the different types of products that the pharmacy carries.
      All `Medications` are initialized with ProductType enum `PRESCRIBED_MEDICINE`. Products can be
      initialized upon instantiation or set with the setter method. Otherwise, all other products
      and items are `UNLABELED` by default.
    - `register/PaymentType` represents the different types of payment a customer can provide to
      complete their transaction at the register. Thus, `PaymentType` is a field implemented
      in `Register`. `NOT_SELECTED` is the default enum and can be changed by either the employee
      operating the register or the customer in future development.
    - `person/EmployeeType` represents the different types of employee the pharmacy can have. By
      default, all employees are `FULL_TIME` employees.

File: src/main/java/genericLinkedList/PrescriptionRequestLog.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        enqueue(prescription);
+    }
+
+    public Prescription getNextPrescriptionRequest(){
+        if (isEmpty()) {
+        LOG.warn("The prescriptionRequestLog is empty");
+        return null;
+    }
+        return dequeue();
+    }
```
Comment (ankochergin) [inline src/main/java/genericLinkedList/PrescriptionRequestLog.java]: formatting

File: src/main/java/misc/BusinessDays.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package misc;
+
+public enum BusinessDays {
```
Comment (ankochergin) [inline src/main/java/misc/BusinessDays.java]:
```suggestion
public enum BusinessDay {
```

File: src/main/java/customLambdaFunctions/NullChecker.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package customLambdaFunctions;
+
+public interface NullChecker<T> {
```
Comment (ankochergin) [inline src/main/java/customLambdaFunctions/NullChecker.java]:
```suggestion
public interface INullChecker<T> {
```
use same naming conventions for functional interfaces

**Review by ankochergin — APPROVED** [review]

**Pull request merged**